errors/errors.go

```diff
@@ -24,16 +28,28 @@ const (
 
 // BoulderError represents internal Boulder errors
 type BoulderError struct {
-	Type   ErrorType
-	Detail string
+	Type      ErrorType
+	Detail    string
+	SubErrors []SubBoulderError
+}
+
+// TODO(@cpu): SubBoulderError type docstring
+type SubBoulderError struct {
+	BoulderError
+	Identifier identifier.ACMEIdentifier
 }
 
-func (be *BoulderError) Error() string {
+func (be BoulderError) Error() string {
 	return be.Detail
 }
 
+func (be *BoulderError) WithSubErrors(subErrs []SubBoulderError) *BoulderError {
+	be.SubErrors = append(be.SubErrors, subErrs...)
+	return be
+}
+
 // New is a convenience function for creating a new BoulderError
-func New(errType ErrorType, msg string, args ...interface{}) error {
+func New(errType ErrorType, msg string, args ...interface{}) *BoulderError {
```

Comment:
Returning `*BoulderError` instead of `error` makes it possible to call `WithSubErrors` on the returned errors from this and the convenience functions. It makes no difference to the callers since `*BoulderError` is an `error`.

Reply:
It does make a difference because of [typed nils](https://dave.cheney.net/2017/08/09/typed-nils-in-go-2). Those are really challenging and painful and I'd like to avoid them at all costs.

Maybe we can forgo `WithSubProblems` for now and let the modest number of call sites that can result in subproblems make their own with struct syntax?

Reply:
We already have to deal with typed nils extensively because of problems. I'm not super convinced.